### DISCUSSION TITLE
docs: Usage with passport js

### DISCUSSION THREAD

**Discussion by ArielGueta** [trpc/trpc#3288] [Q&A] [answered]
### Area of Improvement

Is there any way we can use passport js with trpc? I'm trying to migrate an existing project to trpc, and this is my express code right now:

```ts
const router = express.Router();

passport.use(new GitHubStrategy({
    clientID: environment.GITHUB.CLIENT_ID,
    clientSecret: environment.GITHUB.CLIENT_SECRET,
    callbackURL: `${environment.BASE_URL}/github/callback`,
    scope: ['user:email']
  },
  (accessToken, refreshToken, profile, done) => {
    done(null, profile);
  }
));

router.get('/auth/github', passport.authenticate('github'));

router.get('/github/callback',
  passport.authenticate('github', { session: false }),
  handleAuth);
```

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing the suggested changes!

**Comment by hdwatts** [comment, marked as answer]
First, create a `createContext` method that you can use in your trpc api to ensure the users session is available to trpc middleware

```ts
export const createContext = async (opts) => {

	// Your logic to get the user's session form the request (use opts.req for the request)

	return {
		session
	}
}
```

Then find where you initTRPC on your backend, mine looks like the below:
```ts
const t = initTRPC.context<Context>().create({
	transformer: superjson,
	errorFormatter({ shape }) {
		return shape
	},
})
```

Then create middleware for identifying if your request is auth'd or not. In my case I check for the `user` in the `session`, but that's just how I have it configured

```ts
const isAuthed = t.middleware(({ ctx, next }) => {
	if (!ctx.session?.passport?.user) {
		throw new TRPCError({
			code: "UNAUTHORIZED",
			message: "Not authenticated",
		})
	}

	return next({ ctx })
})
```

Finally, export a procedure that uses the middleware from above (you can also export a publicProcedure that does not run the middleware):

```ts
export const publicProcedure = t.procedure
export const protectedProcedure = t.procedure.use(isAuthed)
```


You can then use that procedure in your router:

```ts
export const postRouter = router({
	all: protectedProcedure.query(({ ctx }) => {
		return ctx.prisma.post.findMany({ orderBy: { createdAt: "desc" } })
	}),
})
```

If `isAuthed` errors, the query never gets called.

**Reply by ArielGueta** [reply]
That's not what I want. Passport provides oauth strategies out of the box.

**Reply by ArielGueta** [reply]
I mean where is the passport configuration?

**Reply by hdwatts** [reply]
What does the passport configuration have to do with TRPC? I wouldn't think TRPC is meant to handle your authentication logic as that's not what it's meant to accomplish. You should maintain your existing express based login routes and just use the resulting cookie/session within TRPC to protect the relevant TRPC routes.

**Reply by ArielGueta** [reply]
So you have a mix of express routes and trpc?

**Reply by hdwatts** [reply]
I do - I have my authentication handlers separate from my trpc endpoint.

I guess you can technically create TRPC handlers and use the vanilla client to execute the query - however due to the different structures between `express` middleware and `trpc` middleware I would think the conversion could be painful.  I haven't tried it but you can possibly include the entire request and response into context and make a wrapper middleware for `passport.authenticate` for your specific routes?

Up to you if it's worth the extra headache, for me I use TRPC for typesafety across my business logic, and want to keep auth separate.
